app/questionnaire/router.py

```diff
@@ -66,44 +66,28 @@ def can_access_hub(self):
     def routing_path(self, section_id, list_item_id=None):
         return self._path_finder.routing_path(section_id, list_item_id)
 
-    def get_next_location_url(self, location, routing_path):
+    def get_next_location_url(self, location, routing_path, return_to_summary=False):
         """
         Get the first incomplete block in section/survey if trying to access the section/survey end,
         and the section/survey is incomplete or gets the next default location if the above is false.
         """
-        current_block_type = self._schema.get_block(location.block_id)["type"]
         last_block_id = routing_path[-1]
-        last_block_type = self._schema.get_block(last_block_id)["type"]
         hub_enabled = self._schema.is_hub_enabled()
+        has_section_summary = self._schema.is_summary_in_section(location.section_id)
 
         if (
-            hub_enabled
-            and location.block_id == last_block_id
-            and self._progress_store.is_section_complete(
-                location.section_id, location.list_item_id
-            )
+            return_to_summary or last_block_id == location.block_id
+        ) and self._progress_store.is_section_complete(
+            location.section_id, location.list_item_id
         ):
-            return url_for(".get_questionnaire")
 
-        # If the section is complete and contains a SectionSummary, return the SectionSummary location
-        if (
-            last_block_type == "SectionSummary"
-            and current_block_type != last_block_type
-            and self._progress_store.is_section_complete(
-                location.section_id, location.list_item_id
-            )
-        ):
-            return url_for(
-                "questionnaire.block",
-                block_id=last_block_id,
-                list_name=routing_path.list_name,
-                list_item_id=routing_path.list_item_id,
-            )
+            if has_section_summary:
+                return self.get_section_summary_url(location)
+            if hub_enabled:
+                return url_for(".get_questionnaire")
 
         if self.is_survey_complete() and not hub_enabled:
-            last_section_id = self._schema.get_section_ids()[-1]
-            last_block_id = self._schema.get_last_block_id_for_section(last_section_id)
-            return Location(section_id=last_section_id, block_id=last_block_id).url()
+            return self.get_last_section_last_block_location().url()
 
         block_id_index = routing_path.index(location.block_id)
         # At end of routing path, so go to next incomplete location
```

Comment:
The logic after this comment is checking to see if the current location is at the end of the routing path, which duplicates some of the logic above. We should look to rationalise this logic about section summary/hub/final summary.